### PR TITLE
Social: Add nextdoor embed block

### PR DESCRIPTION
**Pull request by gmjuhasz**
This changes adds a new block to Jetpack that renders a Nextdoor post.

## Proposed changes:

* Added the JS for the block in the editor
* Added controls to change the URL for a present block
* Added rendering the iframe in PHP based on the attribute

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

https://github.com/orgs/Automattic/projects/733/views/2?filterQuery=&pane=issue&itemId=41283324
## Does this pull request change what data or activity we track or use?

No.
## Testing instructions:

* To see beta blocks add `define( 'JETPACK_BLOCKS_VARIATION', 'beta' );` to your `wp-config.php` file
* Create a new post and test the new Nextdoor blog
  * Typing `/nextdoor` should bring up the empty block
  * Insert a Nextdoor post url ( `https://nextdoor.com/p/<post_id>` ) or embed URL ( `https://nextdoor.com/embed/<post_id>` ), then click embed to turn it into an iframe.
  * Nextdoor has different domains ( `https://nextdoor.co.uk` ) those should work too
  * Using any other string should give an error
  * If you already embedded something into the block, you can change it in the block settings in the editor by clicking on the block
  * Pasting in any of the links above should instantly convert into the block
* Publish the post and test the following
  * Nextdoor blocks with valid URL should be rendered on the post
  * Invalid blocks should not get rendered on the post

| Editor | Published Post |
| ------------- | ------------- |
| CleanShot 2023-10-25 at 15 23 56 png | CleanShot 2023-10-25 at 15 24 20 png |

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/nextdoor/index.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import { createBlock } from '@wordpress/blocks';
+import { __, _x } from '@wordpress/i18n';
+import { getIconColor } from '../../shared/block-icons';
+import attributes from './attributes';
+import edit from './edit';
+import save from './save';
+import './editor.scss';
+import { REGEX, parseUrl } from './utils';
+
+export const name = 'nextdoor';
```
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/nextdoor/index.js]: Given that we're moving towards `block.json` to declare blocks, do you think you could take a similar approach for this new block?

You can take a look at 8447ef5075727869218037724844a6b6ba29b35c to see how the change was done for the Calendly block, for example.

File: projects/plugins/jetpack/extensions/blocks/nextdoor/nextdoor.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+function load_assets( $attr ) {
+	$pattern = '/^http[s]?:\/\/((?:www\.)?nextdoor(?:.*)?\/(?:embed)\/[a-z0-9\/\?=_\-\.\,&%$#\@\!\+]*)/i';
+
+	$url = isset( $attr['url'] ) ? $attr['url'] : null;
+
+	if ( empty( $url ) ) {
+		return;
+	}
+
+	if ( ! preg_match( $pattern, $url ) ) {
```
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/nextdoor/nextdoor.php]: What do you think about switching to `Jetpack_Gutenberg::validate_block_embed_url()` here, for extra safetly?

File: projects/plugins/jetpack/extensions/blocks/nextdoor/utils.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+export const REGEX = /(^|\/\/)(nextdoor\.[^"']*)/i;
+
+const PATH_REGEX = /([^/]+$)/;
+
+const getEmbedUrlFromPostUrl = postUrl => {
```
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/nextdoor/utils.js:5]: Do you think we could have a test for this utility function?

**Review by jeherve — CHANGES_REQUESTED** [review]

File: projects/plugins/jetpack/extensions/blocks/nextdoor/utils.js (hunk 3 above)
Comment (gmjuhasz) [reply projects/plugins/jetpack/extensions/blocks/nextdoor/utils.js:5]: Definitely! I was planning on adding tests overall, just wanted to wait for a round of reviews as this is my first time creating a block, so if there are bigger changes needed I do not write tests for those 👍

File: projects/plugins/jetpack/extensions/blocks/nextdoor/nextdoor.php (hunk 2 above)
Comment (gmjuhasz) [reply projects/plugins/jetpack/extensions/blocks/nextdoor/nextdoor.php]: Addressed in https://github.com/Automattic/jetpack/pull/33751/commits/4d78df79aeac2f9b8b4b9abb19603d1d2520429f

File: projects/plugins/jetpack/extensions/blocks/nextdoor/index.js (hunk 1 above)
Comment (gmjuhasz) [reply projects/plugins/jetpack/extensions/blocks/nextdoor/index.js]: Addressed in https://github.com/Automattic/jetpack/pull/33751/commits/8687cbe77fa99f7200d5ccfee483d50b1e6f9709

File: projects/plugins/jetpack/extensions/blocks/nextdoor/edit.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	const { url } = validatedAttributes;
+	const [ iframeUrl, setIframeUrl ] = useState( url ? parseUrl( url ) : null );
+
+	const [ nextdoorShareUrl, setNextdoorShareUrl ] = useState( iframeUrl );
+
+	const setErrorNotice = () => {
+		noticeOperations.removeAllNotices();
+		noticeOperations.createErrorNotice(
+			__(
+				"Your Nextdoor post couldn't be embedded. Please double check your URL or code.",
```
Comment (manzoorwanijk) [inline projects/plugins/jetpack/extensions/blocks/nextdoor/edit.js]: May be use a generic message?
```suggestion
				"Your Nextdoor post couldn't be embedded. Please double check the URL or code.",
```

File: projects/plugins/jetpack/extensions/blocks/nextdoor/utils.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
+export const REGEX = /(^|\/\/|www.)(nextdoor\.[^"']*)/i;
```
Comment (manzoorwanijk) [inline projects/plugins/jetpack/extensions/blocks/nextdoor/utils.js]: I think that dot after www should be escaped:
```suggestion
export const REGEX = /(^|\/\/|www\.)(nextdoor\.[^"']*)/i;
```

File: projects/plugins/jetpack/extensions/blocks/nextdoor/nextdoor.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ * @return string
+ */
+function load_assets( $attr ) {
+	if ( ! isset( $attr['url'] ) ) {
+		return;
+	}
+
+	$url = Jetpack_Gutenberg::validate_block_embed_url(
+		$attr['url'],
+		array( '/^http[s]?:\/\/((?:www\.)?nextdoor(?:.*)?\/(?:embed)\/[a-z0-9\/\?=_\-\.\,&%$#\@\!\+]*)/i' ),
```
Comment (manzoorwanijk) [inline projects/plugins/jetpack/extensions/blocks/nextdoor/nextdoor.php]: A simplified version of this regext would be to allow any non-whitespace character at the end of the URL
```suggestion
		array( '/^http[s]?:\/\/((?:www\.)?nextdoor(?:.*)?\/(?:embed)\/\S*)/i' ),
```

**Review by manzoorwanijk — DISMISSED** [review]
It works well, it's just that the embed height doesn't work well

Screenshot 2023-10-30 at 12 14 33 PM

**Review by manzoorwanijk — APPROVED** [review]

**Review by jeherve — APPROVED** [review]
This is looking good, nice work! 🥳 

I'd encourage you to merge this one and iterate in follow-up PRs.

**Comment by simison** [comment]
Hi! How does this embed render in:

- RSS feeds
- WP.com reader web / apps
- Emails

**Comment by pablinos** [comment]
> Hi! How does this embed render in:
> 
>     * RSS feeds
> 
>     * WP.com reader web / apps
> 
>     * Emails

In https://github.com/Automattic/jetpack/pull/33931 I've modified it so it'll render as a link to the post if the script can't run to render the embed.

**Pull request merged**